ethcore/src/engines/block_reward.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use ethabi;
+use ethabi::ParamType;
+use ethereum_types::{H160, Address, U256};
+
+use block::ExecutedBlock;
+use error::Error;
+use machine::EthereumMachine;
+
+pub type SystemCall<'a> = FnMut(Address, Vec<u8>) -> Result<Vec<u8>, String> + 'a;
+
+use_contract!(block_reward_contract, "BlockReward", "res/contracts/block_reward.json");
+
+#[repr(u8)]
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum RewardKind {
+	Author = 0,
+	Uncle = 1,
+	EmptyStep = 2,
+}
+
+impl From<RewardKind> for u16 {
+	fn from(reward_kind: RewardKind) -> Self {
+		reward_kind as u16
+	}
+}
+
+pub struct BlockRewardContract {
+	address: Address,
+	block_reward_contract: block_reward_contract::BlockReward,
+}
+
+impl BlockRewardContract {
+	pub fn new(address: Address) -> BlockRewardContract {
+		BlockRewardContract {
+			address,
+			block_reward_contract: block_reward_contract::BlockReward::default(),
+		}
+	}
+
+	pub fn reward(
+		&self,
+		benefactors: &[(Address, RewardKind)],
+		caller: &mut SystemCall,
+	) -> Result<Vec<(Address, U256)>, Error> {
+		let reward = self.block_reward_contract.functions().reward();
+
+		let input = reward.input(
+			benefactors.iter().map(|&(address, _)| H160::from(address)),
+			benefactors.iter().map(|&(_, ref reward_kind)| u16::from(*reward_kind)),
+		);
+
+		let output = caller(self.address, input)
+			.map_err(Into::into)
+			.map_err(::engines::EngineError::FailedSystemCall)?;
+
+		// since this is a non-constant call we can't use ethabi's function output
```

Comment:
is there an issue filed about that?

Reply:
I don't think so, there's also no general way to do it since it involves creating a transaction and the return data is not included in the receipt (we could trace the transaction).

Reply:
The reward contract might have internal state that needs to be updated.

Reply:
which can be used to track total supply if you have a reward contract from the beginning. Or even just to maintain a record of recent rewards if you fork to include one.